src/auxiliary/JSON.cpp

```diff
@@ -0,0 +1,94 @@
+#include "openPMD/auxiliary/JSON.hpp"
+
+#include <vector>
+
+#include "openPMD/config.hpp"
+
+namespace openPMD
+{
+namespace auxiliary
+{
+    TracingJSON::TracingJSON() : TracingJSON( nlohmann::json() )
+    {
+    }
+
+    TracingJSON::TracingJSON( nlohmann::json originalJSON )
+        : m_originalJSON(
+              std::make_shared< nlohmann::json >( std::move( originalJSON ) ) )
+        , m_shadow( std::make_shared< nlohmann::json >() )
+        , m_positionInOriginal( &*m_originalJSON )
+        , m_positionInShadow( &*m_shadow )
+    {
+    }
+
+    nlohmann::json const &
+    TracingJSON::getShadow()
+    {
+        return *m_positionInShadow;
+    }
+
+    nlohmann::json
+    TracingJSON::invertShadow()
+    {
+        nlohmann::json inverted = *m_positionInOriginal;
+        invertShadow( inverted, *m_positionInShadow );
+        return inverted;
+    }
+
+    void
+    TracingJSON::invertShadow(
+        nlohmann::json & result,
+        nlohmann::json & shadow )
```

Comment:
is `shadow` a `const &`?

Reply:
Yes, this method mustn't modify `shadow`. I added the `const`.